mitodl/mitxpro
# Find old vouchers, ensure unique pdf names, add more error logging

#### Pre-Flight checklist


#### What are the relevant tickets?
Fixes #810 
Fixes #811 

#### What's this PR do?
- Creates unique S3/file names for each voucher upload
- Retrieves existing vouchers with the same info instead of always creating new ones
- Adds more error logging if something goes wrong (no matching courses, etc)
- Prevents the same voucher from being uploaded by more than one user

#### How should this be manually tested?
- Set the following .env variables to the same as used on CI:
```
VOUCHER_***
AWS_ACCESS_KEY_ID=
AWS_SECRET_ACCESS_KEY=
AWS_STORAGE_BUCKET_NAME=
```

Also set the following:
```
MITXPRO_USE_S3=True
VOUCHER_COMPANY_ID=1
```

- Create a course run, product, product version, and 100% off single-use coupons by company (id=1) that will match a sample voucher pdf that I can send you.
- Go to `/voucher/upload` and upload the voucher
- You should be able to select the course run and get to the checkout page.  Go ahead and order it.
- Upload the same voucher again.  You should get redirected to the resubmit page.

- Upload another sample voucher that has no matching course run.
- Check the logs, there should be an error about no matching courses found.
- Log in as another user, try to upload the same voucher.  You should get redirected to the resubmit page.
- Check the logs, there should be an error about the voucher being uploaded by a different user
